### ISSUE TITLE
Feature Request: Placeholder support for being set from an Op

### ISSUE THREAD

**Issue by ahundt** [tensorflow/tensorflow#10837]
Update: I thought on this some more, and perhaps if tf.Variable could directly support being set from an op with the transfer happening entirely on the C++ side that might cleanly and simply solve the problem.

Often it is not known immediately what kind of Op will be needed for input, but it would be preferable to specify that detail later without the overhead associated with transporting data in python. 

For example, it would be ideal to be able to define a full model for a task like ImageNet labeling with a LazyOp defined for the Input and Labels. Then later on the output of a `Dataset`, `RecordInput`, or `tf.placeholder` could be supplied to finalize the necessary connection.

This request is based on the comments of @fchollet at https://github.com/fchollet/keras/pull/6928#issuecomment-309552423.

**Comment by tatatodd** [comment]
@fchollet can you comment as to whether this is the TensorFlow feature you were asking for in https://github.com/fchollet/keras/pull/6928#issuecomment-309552423?

**Comment by fchollet** [comment]
In general, the feature in question would be graph rewiring: being able to replace a node with another node in a computation graph, after the fact. The most use case would be to build a graph on top of placeholders, then replace the placeholders with ops that yield data (e.g. TF pipelines, TFRecords, etc). Graph editing.

I don't believe this is currently possible. It would certainly be useful.

**Comment by ahundt** [comment]
Based on the discussion in [Keras Input Tensor API Design Proposal](https://docs.google.com/document/d/1tf2Nl7wor8rmWPUoxfClLuPLQGqvZryegD7K7-1tTe8/edit?usp=sharing) and https://github.com/fchollet/keras/pull/6928#issuecomment-312088200 implementation of this will be required for the ideal API design that supports TFRecords https://github.com/tensorflow/tensorflow/issues/10837 (and any data input op) to work.

**Comment by TimZaman** [comment]
The graph is an append-only data structure. You could hack it up with tf.contrib.graph_editor.

**Comment by ahundt** [comment]
As @TimZaman mentioned, it does look like this might be possible to do manuallly in a pinch with:

- [subgraph.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/graph_editor/subgraph.py)
- [copy_with_input_replacements](https://www.tensorflow.org/api_docs/python/tf/contrib/graph_editor/copy_with_input_replacements) in a pinch, but a real implementation of this feature request is still needed.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by ahundt** [comment]
I think this is still valid and will become especially useful considering the current work on eager execution.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tatatodd** [comment]
@mrry might have some thoughts, as this is related to input data pipelines.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by mrry** [comment]
We're currently mulling over a design for this feature. It seems feasible to extend the `feed_dict` mechanism so that it would be possible to feed a placeholder based on the output of another operation, but it will require some care to ensure that the resulting graph is valid.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mrry** [comment]
I have a prototype implementation that's currently under review.

**Comment by ahundt** [comment]
Awesome! Is there a PR # or is it currently internal?

**Comment by mrry** [comment]
It looks like the initial Python changes landed as part of 9db93a8db8fafee8092b30ae33a26248867e9b26, and the final C++ change landed in 49c6489368ea98feb3259d54a10c6fdfd01caf44. You can play with the private `Session._make_callable_from_options()` API, but be warned that it's still private so it might change. Example code here:

https://github.com/tensorflow/tensorflow/blob/f9c5e71104cb30583127fdc918591cc7604f17ca/tensorflow/python/client/session_test.py#L1374-L1384

The eventual plan is to replace the implementation of `Session.make_callable()` with this new mechanism and expose tensor connections as an option in that API.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ahundt** [comment]
I think this is done & has been integrated into keras even. Thanks!

**Comment by beasteers** [comment]
Wait, sorry to be that person, but I'm confused abt how to do this (with keras specifically, meaning that I'm not trying to touch `sess/feed_dict`). 

Is this just what's happening with `tf.keras.backend.function`? Or is there some way to do `a_placeholder.set_value(tensor)` or `a_placeholder.set_default(tensor)` or something similar
